Fire: Rewrite fire sound code

paramat opened the pull request:
Previous code:
Used looped sounds without attaching them to objects or playing direct
to players.
The looped sounds are not 'stopped' when players leave the area.
These may be causing the bug where sounds are heard at extreme
distances.
Entering a world with already present flames results in silent flames.
Sounds are often played at a large number of points in a 6 node lattice.
A large fire is reported to cause a high load, disabling the sound code
is reported to help this.

New code:
Optional flame sound to not interfere with ambience mods.
Permanent flame now has sound.
For multiple flames, sound is positioned at the centre of all flames, and
has volume determined by flame number.
The original freesound 'large fire' recording was used to create 3 sounds
that play at random for a non-repetitive effect. At low volume it is
suitable for small fires.

Original sound files and sound function (as an empty function) kept
temporarily to reduce disruption.

Reduce gain of flame extinguish sound.
////////////////////////////////////////////////////

See #1112 and #1385 for discussion.
The existing sound code has serious problems and bugs so must be removed for release.

SmallJoker commented on mods/fire/init.lua:
```diff
@@ -151,72 +161,92 @@ minetest.override_item("default:coalblock", {
+        if obj and obj:is_player() then
+            fire.update_player_sound(obj)
+        end
+    end
```
Holy spaces to be removed please.

SmallJoker commented on mods/fire/init.lua:
```diff
@@ -151,72 +161,92 @@ minetest.override_item("default:coalblock", {
+		)
+		-- Store sound handle for this player
+		if handle then
+			handles[player_name] = handle
```
Even this is just one number in this table, the handle should be deleted when the player leaves.

paramat replied:
Ok.

paramat replied:
Done, also does 'sound stop'.

SmallJoker commented on mods/fire/init.lua:
```diff
@@ -151,72 +161,92 @@ minetest.override_item("default:coalblock", {
+	local ppos = player:getpos()
+	local fpos, num = minetest.find_nodes_in_area(
+		{x = ppos.x - radius, y = ppos.y - radius, z = ppos.z - radius},
+		{x = ppos.x + radius, y = ppos.y + radius, z = ppos.z + radius},
```
`vector.add` and `vector.subtract` were added for this purpose.

kaeza commented:
While you're at it, can you reset the handle to nil on player leave?

paramat commented:
Updated with all requests, also do 'sound stop' on player leave, in case that is helpful.

paramat commented:
Has error, fixing.

paramat commented:
Tuned sound gain and fixed error.

Fixer-007 commented:
Will check

paramat commented:
Everything you mention is for a reason, these becomes clear once you work on this, i'll try to explain soon. The reaction time we discussed before, it's a balance against the constant load of searching around players, it's less critical now because any add or remove of flames causes an instant update.

paramat commented:
However the code that chooses a fairly 'central' position to place the sound could be improved.
Currently it just chooses the midpoint of the flame position table, which is often not particularly central, and jumps around a lot as flames appear and disappear.
Instead i could search through the table of flame positions and find the min and max points of the cuboid containing all flames, then choose a point halfway betwen these.
Also i may be able to re-add the 'fade-in' 'fade-out' samples from the previous implementation.

paramat commented:
Updated with method described in previous comment.
However there is a problem. I added `print()` to see how often the instant updated are called, and in a large forest fire these come in large multiples of 4-12 at once, every few seconds, because each ABM cycle is adding or removing multiple flames at once. I'm trying to work out a way to rate-limit these or just trigger one instant update per multiple batch.

paramat commented:
Essentially, implementing ambience sounds in lua like this is either:
Very intensive and barely acceptable.
or,
Slightly less intensive but very basic and unresponsive (my previous code) such as 3 seconds to update sounds when moving or placing/removing flames.
It may be better, and actually less embarrassing, to have no fire sounds at all and move this type of code into optional mods. Ambience sounds really need new engine code to be done acceptably.

sofar commented:
We talked about this in IRC a bit. The gist of it is that I've been cautious about rewriting this code because it opens up a can of worms.

I can see that there will be a decent solution, but it will be costly in terms of network packets, or it will be cheap, and sound bad.

I've stayed out of this discussion for a bit to see if it would actually go somewhere, but it seems that @paramat is reaching the same conclusion that I have: solving this problem is going to be complex, and not something we can do in just a few days.

I have no solution atm. Removal of the current code is crude, as well.

paramat commented:
Updated to simple version. There is up to a 3s lag though.
The instant updates are removed due to the problems they caused, now just uses the 3 second cycle of updates.
Docs still need updating for the sound setting.

paramat commented:
A single flame may seem quiet, but it's chosen to be a gentle sound suitable for a campfire or a large torch above a coalblock.

paramat commented:
![screenshot_20161206_055611](https://cloud.githubusercontent.com/assets/3686677/20914659/05d42a52-bb79-11e6-82dc-42efa278e036.png)

You can test the maximum possible volume with this pattern and stand in the hole.

paramat commented:
Updated.
Made necessary changes to conf.example and settingtypes.
Original sound files and sound function (as an empty function) kept temporarily to reduce disruption.
Reduce gain of flame extinguish sound.

paramat commented:
Tested.
Settings in .conf and using advanced settings menu are tested.
One of the fire extinguish sounds is louder than the others, and since several flames can be extinguished at once the sound can get very loud. It seems balanced now.

sofar commented:
Good. Definitely lightweight. A bit odd when there's 2 distinct fire spots and you're in between, but given the absolute minimal resources this uses I think it's acceptable.

Might want to consider doing a slightly larger search radius, or tuning max hear distance to the radius value, so that attenuation code in openal can account for that.

:+1: irregardless comments.

paramat commented:
Yes radius 8 could be good, 6 feels a bit short.
I used hear distance 32 to not have sound fall so rapidly with distance, otherwise it tends to become too quiet when just a few nodes away.

paramat commented:
Updated radius to 8.

paramat commented:
fed2151d708b79b86bed7f112c2356966f2b4500

Fixer-007 commented:
I've tested merged version of this with mod profiler:
```
Values below show absolute/relative times spend per server step by the instrumented function.
A total of 9067 samples were taken


 instrumentation                                         |    min µs |    max µs |    avg µs | min % | max % | avg %
-------------------------------------------------------- | --------- | --------- | --------- | ----- | ----- | ------
 sfinv:                                                  |         0 |     15940 |        22 |   0.0 |  99.8 |   0.2
  - on_player_receive_fields[1] .......................  |         5 |     15940 |      7972 |  16.1 |  99.8 |  58.0
 flowers:                                                |         0 |       441 |         0 |   0.0 |  93.4 |   0.1
  - ABM 'Mushroom spread' .............................  |         3 |        19 |         7 |   0.1 |   2.2 |   0.8
  - ABM 'Flower spread' ...............................  |         3 |       441 |       153 |   0.0 |  93.4 |  29.9
 nyancat:                                                |         0 |        68 |         0 |   0.0 |  68.7 |   0.0
  - on_generated[1] ...................................  |         2 |        68 |        16 |   0.1 |  68.7 |  29.6
 beds:                                                   |         0 |         1 |         0 |   0.0 |   3.2 |   0.0
  - on_player_receive_fields[1] .......................  |         1 |         1 |         1 |   3.2 |   3.2 |   3.2
 fire:                                                   |         0 |     21553 |        71 |   0.0 |  99.9 |   4.7
  - globalstep[1] .....................................  |         1 |      2029 |         8 |   0.0 |  99.3 |   8.0
  - ABM 'Ignite flame' ................................  |       271 |     11182 |      5341 |  10.5 |  99.9 |  65.8
  - ABM 'Remove flammable nodes' ......................  |        76 |     13329 |      5621 |   1.8 |  99.9 |  64.7
 default:                                                |         4 |      7337 |       100 |   0.1 | 100.0 |  95.1
  - globalstep[1] .....................................  |         4 |      2945 |        18 |   0.1 | 100.0 |  88.3
  - ABM 'Leaf decay' ..................................  |       228 |      7134 |      4148 |  10.8 |  99.8 |  82.0
  - get_staticdata[1] .................................  |        15 |        28 |        22 |   0.4 |   0.6 |   0.5
  - on_step[1] ........................................  |         3 |       188 |        14 |   0.1 |  91.7 |  42.6
  - on_activate[1] ....................................  |        30 |        61 |        46 |   0.7 |   1.2 |   1.0
  - ABM 'Grass spread' ................................  |         2 |       245 |        80 |   0.1 |  94.6 |  43.0
  - on_player_receive_fields[1] .......................  |         1 |         1 |         1 |   3.2 |   3.2 |   3.2
  - ABM 'Grass covered' ...............................  |        12 |       388 |       161 |   0.3 |  37.2 |   4.5
-------------------------------------------------------- | --------- | --------- | --------- | ----- | ----- | ------
 total:                                                  |         4 |     25234 |       173 | 100.0 | 100.0 | 100.0
```
My impressions: pretty simple, mostly works, in complex fire situations may give weird source positions, also somewhat bumpy sound change when moving through large forest fire (fadeout - sudden volume increase - fade out).

paramat commented:
Thanks, useful, looks lightweight.
